app/src/main/java/org/mozilla/firefox/vpn/user/data/ReferralManager.kt

```diff
@@ -0,0 +1,64 @@
+package org.mozilla.firefox.vpn.user.data
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Log
+import com.android.installreferrer.api.InstallReferrerClient
+import com.android.installreferrer.api.InstallReferrerStateListener
+import com.android.installreferrer.api.ReferrerDetails
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class ReferralManager(context: Context, private val prefs: SharedPreferences) {
+
+    companion object {
+        private const val TAG = "ReferralManager"
+        private const val SIR_PREF_REFERRAL = "sir_pref_referral"
+        private const val DEFAULT_REFERRAL = "default"
+    }
+
+    private var referrerClient: InstallReferrerClient =
+        InstallReferrerClient.newBuilder(context).build()
+
+
+    suspend fun getUserReferral() = suspendCoroutine<String> { cont ->
```

Comment:
Just a reminder, make sure no multiple calls to cont.resume() will happen, otherwise an IllegalStateException will be thrown. You can reference CustomTamUtil#initCustomService(), where multiple service connection callback is called, though the official document guarantee it will be called only once

Reply:
fixed.